docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,113 @@
+- Allow the addition of new profiles, device resources, device commands, etc. **always**.  Additions may be accomplished via POST or PATCH depending on the change and API.
+
+See [Proposed Design section](#proposed-design) section below for specific metadata API changes.
+
+While not explicitly covered in this ADR, a device is not allowed to be deleted when it is associated to an event or reading.
+
+### Profile Changes
+
+The table below outlines the current elements of the EdgeX device profile, and the circumstances allowed for change (accomplished via PUT update API call - whether that is add, update or delete elements where applicable).
+
+| Device Profile Property | Type | Allowed Change | Description/Comments |
+|-------------------------|------|----------------|----------------------|
+| *General Properties* ||||
+| Id | string | Never | This uuid is generated by EdgeX and uniquely identifies the profile. |
+| Name | string | Never | This required property must be unique to the EdgeX instance and cannot be changed.  The profile name can never be empty and must be unique across the entire EdgeX instance. |
+| Manufacturer | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Description | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Model | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Labels | string array | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include adding, removing or modifying any of the strings in the array. |
+| Device Resources | array | Situational | New device resources can always be added. Existing device resources can only be removed from the array when the device resource has not yet been associated to any reading or event.|
+| Device Commands | array | Situational | New device commands can always be added.  Existing device commands can only be removed from the array when the device command has not yet been associated to a reading or event.  Please see the device command properties for rules on modifying device commands.Device commands are optional and so the array can always be empty. |
+| *Device Resource Properties* |||
+| Description | string | Anytime | As this property does not affect any functionality, this property can be changed at any time – to include setting the string to an empty string. |
+| Name | string | Situational | When the device resource is not associated to any reading or event, this field can be modified.  It can never be blank and must be unique across the entire profile. |
```

Comment:
@cloudxxx8 , how does changing the "key" impact anything is there are not ant references to it?

Reply:
when you send out a PATCH request, it might look like
PATCH /api/v2/deviceprofile/name/{profile name}/deviceresource
```
{
  "name": "resourceA",
  "description":"the new one"
}
```

The implementation need to query the "resourceA" device resource and modify the "description".
If the users want to modify the name, how do we know which device resource should be modified?
We encountered the same situation, and we decided to make the `name` cannot be updated in all kinds of objects.

The workaround is to put the device resource original name to the URL, but it is not consistent to other APIs

Reply:
Changed to Never change based on Cloud's notes above.